src/power_grid_model/_utils.py

```diff
@@ -151,7 +151,9 @@ def get_batch_size(batch_data: BatchComponentData) -> int:
         return indptr.size - 1
 
     data_to_check = (
-        next(iter(cast(DenseBatchColumnarData, batch_data).values())) if is_columnar(batch_data) else batch_data
+        next(iter(cast(DenseBatchColumnarData, batch_data).values()))
+        if is_columnar(batch_data)
+        else cast(DenseBatchArray, batch_data)
```

Comment:
so many of these kinds of statements. it may be a good idea to use a functional approach instead, e.g.

```py
def apply_row_based_or_columnar(
    data,
    *,
    row_based_func: Callable | None = None,
    columnar_func: Callable | None = None
):
    if is_columnar(data):
        if columnar_func is not None:
            columnar_func(data)
    elif row_based_func is not None:
        row_based_func(data)
```
we'll probably have to do some typevar magic to mimic the appropriate template behavior though, but it should get rid of a lot of the `cast`s